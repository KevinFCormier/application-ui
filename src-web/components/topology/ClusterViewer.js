/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import * as d3 from 'd3'
import resources from '../../../lib/shared/resources'
import config from '../../../lib/shared/config'
import LayoutHelper from './layoutHelper'
import LinkHelper from './linkHelper'
import NodeHelper from './nodeHelper'
import msgs from '../../../nls/platform.properties'
import _ from 'lodash'


resources(() => {
  require('../../../scss/cluster-viewer.scss')
  require('../../../scss/topology-link.scss')
  require('../../../scss/topology-node.scss')
})

var currentZoom = {x:0, y:0, k:1}

class ClusterViewer extends React.Component {

  static propTypes = {
    id: PropTypes.string,
    links: PropTypes.arrayOf(PropTypes.shape({
      source: PropTypes.any,
      target: PropTypes.any,
      label: PropTypes.string,
      type: PropTypes.string,
    })),
    name: PropTypes.string,
    nodes: PropTypes.arrayOf(PropTypes.shape({
      cluster: PropTypes.string,
      uid: PropTypes.string.isRequired,
      type: PropTypes.string,
      name: PropTypes.string,
    })),
    onSelectedNodeChange: PropTypes.func
  }

  constructor (props) {
    super(props)

    this.state = { links: props.links, nodes: props.nodes }

    this.setContainerRef = elem => {
      this.containerRef = elem
    }
    this.layoutHelper = new LayoutHelper()
  }

  componentDidMount() {
    this.generateDiagram()
  }

  componentDidUpdate(){
    this.generateDiagram()
  }

  shouldComponentUpdate(nextProps, nextState){
    return !_.isEqual(this.state.nodes.map(n => n.id), nextState.nodes.map(n => n.id))
    || !_.isEqual(this.state.links.map(l => l.uid), nextState.links.map(l => l.uid))
  }

  componentWillReceiveProps(nextProps){
    let nodes = _.cloneDeep(nextProps.nodes)
    nodes = nodes.map(node => this.state.nodes.find(n => n.uid === node.uid) || node)
    let links = _.cloneDeep(nextProps.links)
    links = links.map(link => this.state.links.find(l => l.uid === link.uid) || link)
    this.setState({ links, nodes })
  }

  render() {
    const { id, name } = this.props
    const { locale } = this.context
    return (
      <div className="clusterViewerDiagram" ref={this.setContainerRef} >
        <div className='clusterViewerTitle'>
          {msgs.get('cluster.name', [name], locale)}
        </div>
        <div className='clusterViewerContainer'>
          <svg id={name+id} className="topologyDiagram" />
          <input type='image' alt='zoom-in' className='zoom-in'
            onClick={this.handleZoomIn} src={`${config.contextPath}/graphics/zoom-in.svg`} />
          <input type='image' alt='zoom-out' className='zoom-out'
            onClick={this.handleZoomOut} src={`${config.contextPath}/graphics/zoom-out.svg`} />
          <input type='image' alt='zoom-target' className='zoom-target'
            onClick={this.handleTarget} src={`${config.contextPath}/graphics/zoom-center.svg`} />
        </div>
      </div>
    )
  }

  handleNodeClick = (node) => {
    this.props.onSelectedNodeChange(node.uid)
    d3.event.stopPropagation()
  }

  generateDiagram() {
    if (!this.containerRef) {
      return
    }

    if (!this.svg) {
      const {id, name} = this.props
      this.svg = d3.select('#'+name+id)
      this.svg.append('g').attr('class', 'clusters')
      this.svg.append('g').attr('class', 'links') // Links must be added before nodes, so nodes are painted on top.
      this.svg.append('g').attr('class', 'nodes')
      this.svg.on('click', this.props.onSelectedNodeChange)
    }

    // consolidate nodes/filter links/add layout data to each element
    const {nodes, links} = this.state
    this.layoutBBox = this.layoutHelper.layout(nodes, links, ()=>{

      // resize diagram to fit all the nodes
      this.zoomFit()

      // Create or refresh the nodes in the diagram.
      const transition = d3.transition()
        .duration(400)
        .ease(d3.easeSinOut)

      // Create or refresh the links in the diagram.
      const linkHelper = new LinkHelper(this.svg, links)
      linkHelper.removeOldLinksFromDiagram()
      linkHelper.addLinksToDiagram(currentZoom)
      linkHelper.moveLinks(transition)

      const nodeHelper = new NodeHelper(this.svg, nodes, linkHelper)
      nodeHelper.removeOldNodesFromDiagram()
      nodeHelper.addNodesToDiagram(currentZoom, this.handleNodeClick)
      nodeHelper.moveNodes(transition)

    })

    // Add zoom feature to diagram
    this.svg.call(this.getSvgSpace())

  }

  getSvgSpace(){
    const svgSpace = d3.zoom()
      .scaleExtent([ 0.25, 4 ])
      .on('zoom', () => {
        currentZoom = d3.event.transform
        const {id, name} = this.props
        const svg = d3.select('#'+name+id)
        svg.select('g.nodes').selectAll('g.node')
          .attr('transform', d3.event.transform)
        svg.select('g.links').selectAll('g.link')
          .attr('transform', d3.event.transform)
      })
    return svgSpace
  }

  handleZoomIn = () => {
    this.getSvgSpace().scaleBy(this.svg, 1.3)
  }

  handleZoomOut = () => {
    this.getSvgSpace().scaleBy(this.svg, 1 / 1.3)
  }

  handleTarget = () => {
    this.zoomFit()
  }

  zoomFit = () => {
    const {width, height} = this.layoutBBox
    if (width && height) {
      const root = this.svg.select('g.nodes')
      const parent = root.node().parentElement
      const fullWidth = parent.clientWidth
      const fullHeight = parent.clientHeight
      const scale = Math.min( 1, .99 / Math.max(width / fullWidth, height / fullHeight))
      this.getSvgSpace().translateTo(this.svg, width/2, height/2)
      this.getSvgSpace().scaleTo(this.svg, scale)
    }
  }
}

export default ClusterViewer
