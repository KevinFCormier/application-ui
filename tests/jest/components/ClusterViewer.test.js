/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import ClusterViewer from '../../../src-web/components/topology/ClusterViewer'

describe('ClusterViewer no components', () => {
  const mockData = {
    clusters: [],
    nodes:[],
    links:[],
    onSelectedNodeChange: jest.fn()
  }
  it('renders as expected', () => {
    const component = renderer.create(
      <ClusterViewer
        nodes={mockData.nodes}
        links={mockData.links}
        onSelectedNodeChange={mockData.onSelectedNodeChange}
        selectedNodeId={mockData.selectedNodeId}
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

const mockData = {
  cluster: {
    id: 'cluster_id',
    name: 'name'
  },
  nodes:[
    {
      'id': '1',
      'uid': '1',
      'name': '1',
      'cluster': 'cluster_id',
      'type': 'service',
      'namespace': 'default',
      'topology': 'cluster',
      '__typename': 'Resource'
    },
    {
      'id': '2',
      'uid': '2',
      'name': '2',
      'cluster': 'cluster_id',
      'type': 'service',
      'namespace': 'default',
      'topology': 'service',
      '__typename': 'Resource'
    },
    {
      'id': '3',
      'uid': '3',
      'name': '3',
      'cluster': 'cluster_id',
      'type': 'service',
      'namespace': 'default',
      'topology': 'service',
      '__typename': 'Resource'
    }],
  links:[
    {
      source: 1,
      target: 2,
      label: 'link',
      type: 'link'
    },
    {
      source: 2,
      target: 3,
      label: 'link',
      type: 'link'
    }
  ],
  onSelectedNodeChange: jest.fn()
}

describe('ClusterViewer 3 components', () => {
  it('renders as expected', () => {
    const component = shallow(
      <ClusterViewer
        id={mockData.cluster.id}
        name={mockData.cluster.name}
        nodes={mockData.nodes}
        links={mockData.links}
        onSelectedNodeChange={mockData.onSelectedNodeChange}
        selectedNodeId={mockData.selectedNodeId}
      />
    )
    expect(component).toMatchSnapshot()
  })
})
