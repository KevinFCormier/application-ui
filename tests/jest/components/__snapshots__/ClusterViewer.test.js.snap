// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterViewer 3 components renders as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ClusterViewer
    id="cluster_id"
    links={
      Array [
        Object {
          "label": "link",
          "source": 1,
          "target": 2,
          "type": "link",
        },
        Object {
          "label": "link",
          "source": 2,
          "target": 3,
          "type": "link",
        },
      ]
    }
    name="name"
    nodes={
      Array [
        Object {
          "__typename": "Resource",
          "cluster": "cluster_id",
          "id": "1",
          "name": "1",
          "namespace": "default",
          "topology": "cluster",
          "type": "service",
          "uid": "1",
        },
        Object {
          "__typename": "Resource",
          "cluster": "cluster_id",
          "id": "2",
          "name": "2",
          "namespace": "default",
          "topology": "service",
          "type": "service",
          "uid": "2",
        },
        Object {
          "__typename": "Resource",
          "cluster": "cluster_id",
          "id": "3",
          "name": "3",
          "namespace": "default",
          "topology": "service",
          "type": "service",
          "uid": "3",
        },
      ]
    }
    onSelectedNodeChange={[MockFunction]}
    selectedNodeId={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="clusterViewerTitle"
        >
          Cluster: {0}
        </div>,
        <div
          className="clusterViewerContainer"
        >
          <svg
            className="topologyDiagram"
            id="namecluster_id"
          />
          <input
            alt="zoom-in"
            className="zoom-in"
            onClick={[Function]}
            src="undefined/graphics/zoom-in.svg"
            type="image"
          />
          <input
            alt="zoom-out"
            className="zoom-out"
            onClick={[Function]}
            src="undefined/graphics/zoom-out.svg"
            type="image"
          />
          <input
            alt="zoom-target"
            className="zoom-target"
            onClick={[Function]}
            src="undefined/graphics/zoom-center.svg"
            type="image"
          />
        </div>,
      ],
      "className": "clusterViewerDiagram",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Cluster: {0}",
          "className": "clusterViewerTitle",
        },
        "ref": null,
        "rendered": "Cluster: {0}",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <svg
              className="topologyDiagram"
              id="namecluster_id"
            />,
            <input
              alt="zoom-in"
              className="zoom-in"
              onClick={[Function]}
              src="undefined/graphics/zoom-in.svg"
              type="image"
            />,
            <input
              alt="zoom-out"
              className="zoom-out"
              onClick={[Function]}
              src="undefined/graphics/zoom-out.svg"
              type="image"
            />,
            <input
              alt="zoom-target"
              className="zoom-target"
              onClick={[Function]}
              src="undefined/graphics/zoom-center.svg"
              type="image"
            />,
          ],
          "className": "clusterViewerContainer",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "topologyDiagram",
              "id": "namecluster_id",
            },
            "ref": null,
            "rendered": null,
            "type": "svg",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "zoom-in",
              "className": "zoom-in",
              "onClick": [Function],
              "src": "undefined/graphics/zoom-in.svg",
              "type": "image",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "zoom-out",
              "className": "zoom-out",
              "onClick": [Function],
              "src": "undefined/graphics/zoom-out.svg",
              "type": "image",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "zoom-target",
              "className": "zoom-target",
              "onClick": [Function],
              "src": "undefined/graphics/zoom-center.svg",
              "type": "image",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="clusterViewerTitle"
          >
            Cluster: {0}
          </div>,
          <div
            className="clusterViewerContainer"
          >
            <svg
              className="topologyDiagram"
              id="namecluster_id"
            />
            <input
              alt="zoom-in"
              className="zoom-in"
              onClick={[Function]}
              src="undefined/graphics/zoom-in.svg"
              type="image"
            />
            <input
              alt="zoom-out"
              className="zoom-out"
              onClick={[Function]}
              src="undefined/graphics/zoom-out.svg"
              type="image"
            />
            <input
              alt="zoom-target"
              className="zoom-target"
              onClick={[Function]}
              src="undefined/graphics/zoom-center.svg"
              type="image"
            />
          </div>,
        ],
        "className": "clusterViewerDiagram",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Cluster: {0}",
            "className": "clusterViewerTitle",
          },
          "ref": null,
          "rendered": "Cluster: {0}",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <svg
                className="topologyDiagram"
                id="namecluster_id"
              />,
              <input
                alt="zoom-in"
                className="zoom-in"
                onClick={[Function]}
                src="undefined/graphics/zoom-in.svg"
                type="image"
              />,
              <input
                alt="zoom-out"
                className="zoom-out"
                onClick={[Function]}
                src="undefined/graphics/zoom-out.svg"
                type="image"
              />,
              <input
                alt="zoom-target"
                className="zoom-target"
                onClick={[Function]}
                src="undefined/graphics/zoom-center.svg"
                type="image"
              />,
            ],
            "className": "clusterViewerContainer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "topologyDiagram",
                "id": "namecluster_id",
              },
              "ref": null,
              "rendered": null,
              "type": "svg",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "zoom-in",
                "className": "zoom-in",
                "onClick": [Function],
                "src": "undefined/graphics/zoom-in.svg",
                "type": "image",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "zoom-out",
                "className": "zoom-out",
                "onClick": [Function],
                "src": "undefined/graphics/zoom-out.svg",
                "type": "image",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "zoom-target",
                "className": "zoom-target",
                "onClick": [Function],
                "src": "undefined/graphics/zoom-center.svg",
                "type": "image",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`ClusterViewer no components renders as expected 1`] = `
<div
  className="clusterViewerDiagram"
>
  <div
    className="clusterViewerTitle"
  >
    Cluster: {0}
  </div>
  <div
    className="clusterViewerContainer"
  >
    <svg
      className="topologyDiagram"
      id={NaN}
    />
    <input
      alt="zoom-in"
      className="zoom-in"
      onClick={[Function]}
      src="undefined/graphics/zoom-in.svg"
      type="image"
    />
    <input
      alt="zoom-out"
      className="zoom-out"
      onClick={[Function]}
      src="undefined/graphics/zoom-out.svg"
      type="image"
    />
    <input
      alt="zoom-target"
      className="zoom-target"
      onClick={[Function]}
      src="undefined/graphics/zoom-center.svg"
      type="image"
    />
  </div>
</div>
`;
